### ISSUE TITLE
dotnet tools X net5.0

### ISSUE THREAD

**Issue by wli3** [dotnet/sdk#12055]
What happens if a dotent tools targeting net5.0-win and want to call Windows specific API? Say access Bluetooth to help debugging a device.

Currently the target framework selection is choosing the most compatible TFM according to the TFM SDK is built on. In this case, SDK will be built on net5.0, and it won't choose net5.0-win.

**Comment by wli3** [comment]
cc @KathleenDollard

**Comment by wli3** [comment]
at the same time I don't think we have `net5.0-linux` and `net5.0-mac`.

**Comment by wli3** [comment]
my opinion is this is a low priority. I don't think there are many tools want to access platform specific API. We could ship without this support and see the reaction.

**Comment by marcpopMSFT** [comment]
https://github.com/dotnet/runtime/issues/46128

**Comment by tthiery** [comment]
It would be lovely if this makes it into the .NET 6 release ;). As in the other issue described, I have exactly the bluetooth case described in the original description.

**Comment by guillaume86** [comment]
I would also use this, my use case is showing Toast notifications if the platform allows it.

**Comment by JanKotschenreuther** [comment]
We also got use cases.

For example we got cli tools which got large lists of options as input.
We actually build UI tools to help managing those lists, keeping track of and changing all options and their states.

**Comment by lindexi** [comment]
How can I ignore the NETSDK1146 error?

**Comment by lindexi** [comment]
> How can I ignore the NETSDK1146 error?

One of the hack code:

```xml
    <Target Name="HackBeforePackToolValidation" BeforeTargets="_PackToolValidation">
        <PropertyGroup>
            <TargetPlatformIdentifier></TargetPlatformIdentifier>
        </PropertyGroup>
    </Target>

    <Target Name="HackAfterPackToolValidation" AfterTargets="_PackToolValidation">
        <PropertyGroup>
            <TargetPlatformIdentifier>Windows</TargetPlatformIdentifier>
        </PropertyGroup>
    </Target>
```

**Comment by LazaroOnline** [comment]
> ```
>     <Target Name="HackBeforePackToolValidation" BeforeTargets="_PackToolValidation">
>         <PropertyGroup>
>             <TargetPlatformIdentifier></TargetPlatformIdentifier>
>         </PropertyGroup>
>     </Target>
> 
>     <Target Name="HackAfterPackToolValidation" AfterTargets="_PackToolValidation">
>         <PropertyGroup>
>             <TargetPlatformIdentifier>Windows</TargetPlatformIdentifier>
>         </PropertyGroup>
>     </Target>
> ```

Pardon my ignorance, but where should I put that code?
I tried in csproj like this:
```xml
<Project Sdk="Microsoft.NET.Sdk">
  ...
  <Target Name="HackBeforePackToolValidation" BeforeTargets="_PackToolValidation">
    ....
  </Target>
</Project>
```
but didn't work. It throws an error "MSB4232	Items that are outside Target elements must have one of the following operations: Include, Update, or Remove."

So I suppose I need to create some other file related to msbuild, but I'm not familiar with that process.
Please, can someone explain how to add that Target hack code?

**Comment by LazaroOnline** [comment]
@KalleOlaviNiemitalo  yes, I also tried placint it into an ItemGroup and had the same error message.
Here is a morde detailed example of the csproj file:
```xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>net7.0-windows</TargetFramework>
    <UseWindowsForms>true</UseWindowsForms>
    <PackAsTool>True</PackAsTool>
  </PropertyGroup>

  <ItemGroup>
    <Content Include="Icon.ico" />
  </ItemGroup>

  <ItemGroup>
    <None Include="..\Icon.png">
      <Pack>True</Pack>
      <PackagePath>\</PackagePath>
    </None>
    <None Include="..\README.md">
      <Pack>True</Pack>
      <PackagePath>\</PackagePath>
    </None>
  </ItemGroup>

  <ItemGroup>
    <PackageReference Include="Microsoft.Extensions.Configuration" Version="7.0.0" />
  </ItemGroup>

  <ItemGroup>
    <Target Name="HackBeforePackToolValidation" BeforeTargets="_PackToolValidation">
      <PropertyGroup>
        <TargetPlatformIdentifier></TargetPlatformIdentifier>
      </PropertyGroup>
    </Target>

    <Target Name="HackAfterPackToolValidation" AfterTargets="_PackToolValidation">
      <PropertyGroup>
        <TargetPlatformIdentifier>Windows</TargetPlatformIdentifier>
      </PropertyGroup>
    </Target>
  </ItemGroup>

</Project>
```
Error:
Items that are outside Target elements must have one of the following operations: Include, Update, or Remove.

**Comment by LazaroOnline** [comment]
@KalleOlaviNiemitalo  True, I got it to work without the ItemGroup, however VisualStudio seems to have some stability problems handling these manual changes, it intermittently complains about the csproj.
I will edit my example above to leave a working example so it is less confusing for others.

**Comment by Anduin2017** [comment]
> > How can I ignore the NETSDK1146 error?
> 
> One of the hack code:
> 
> ```
>     <Target Name="HackBeforePackToolValidation" BeforeTargets="_PackToolValidation">
>         <PropertyGroup>
>             <TargetPlatformIdentifier></TargetPlatformIdentifier>
>         </PropertyGroup>
>     </Target>
> 
>     <Target Name="HackAfterPackToolValidation" AfterTargets="_PackToolValidation">
>         <PropertyGroup>
>             <TargetPlatformIdentifier>Windows</TargetPlatformIdentifier>
>         </PropertyGroup>
>     </Target>
> ```

I tired adding this hack.

However when I was installing it, I got an error:

```bash
$ dotnet tool install MyApp--global
Tool 'xxxxxxxxxxxxxxxx.xxxxx' failed to update due to the following:
The settings file in the tool's NuGet package is invalid: Settings file 'DotnetToolSettings.xml' was not found in the package.
Tool 'xxxxxxxxxxxxxxxx.xxxxx' failed to install. Contact the tool author for assistance.
```

**Comment by lindexi** [comment]
@Anduin2017 Update the hack csproj code for .NET 8.0 or greater

```xml
  <Target Name="HackBeforePackToolValidation" BeforeTargets="_PackToolValidation">
    <PropertyGroup>
      <TargetPlatformIdentifier></TargetPlatformIdentifier>
      <TargetPlatformMoniker></TargetPlatformMoniker>
    </PropertyGroup>
  </Target>

  <Target Name="HackAfterPackToolValidation" AfterTargets="_PackToolValidation" BeforeTargets="PackTool">
    <PropertyGroup>
      <TargetPlatformIdentifier>Windows</TargetPlatformIdentifier>
    </PropertyGroup>
  </Target>
```

All the csproj code:

```xml
<Project Sdk="Microsoft.NET.Sdk.WindowsDesktop">

  <PropertyGroup>
    <OutputType>WinExe</OutputType>
    <TargetFramework>net8.0-windows</TargetFramework>
    <UseWPF>true</UseWPF>
    <PackAsTool>true</PackAsTool>
    <ToolCommandName>lindexi</ToolCommandName>
  </PropertyGroup>

  <Target Name="HackBeforePackToolValidation" BeforeTargets="_PackToolValidation">
    <PropertyGroup>
      <TargetPlatformIdentifier></TargetPlatformIdentifier>
      <TargetPlatformMoniker></TargetPlatformMoniker>
    </PropertyGroup>
  </Target>

  <Target Name="HackAfterPackToolValidation" AfterTargets="_PackToolValidation" BeforeTargets="PackTool">
    <PropertyGroup>
      <TargetPlatformIdentifier>Windows</TargetPlatformIdentifier>
    </PropertyGroup>
  </Target>

</Project>
```

And you can find all my code in https://github.com/lindexi/lindexi_gd/blob/fdd3959ce65cf2ebeead9706c04901caafb1f89c/WinemwhajallawLigawakuja/WinemwhajallawLigawakuja.csproj

**Comment by Anduin2017** [comment]
Thank you @lindexi ! I tried your hack an it works! Now I can distribute my WPF app via dotnet tool!

**Comment by baronfel** [comment]
It's now possible to create RID-specific tools starting in .NET 10p6, and such tools are created via `Publish` instead of via `Build`. This may mean that it is possible to create tools that bind to the Windows APIs you need more easily now. Feel free to check out early examples over at https://github.com/baronfel/multi-rid-tool and see if the scenario is unblocked.
